codec-http2/src/main/java/io/netty/handler/codec/http2/PriorityStreamByteAssigner.java

```diff
@@ -0,0 +1,437 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License, version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at:
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.netty.handler.codec.http2;
+
+import static io.netty.util.internal.ObjectUtil.checkNotNull;
+import static java.lang.Math.max;
+import static java.lang.Math.min;
+
+import java.util.Arrays;
+
+/**
+ * A {@link StreamByteAssigner} that implements the HTTP/2 priority tree algorithm for assigning
+ * bytes for all streams in the connection.
+ */
+public final class PriorityStreamByteAssigner implements StreamByteAssigner {
```

Comment:
In general, I don't subscribe to the idea that RFC terminology necessarily maps to good names in the software design. When there are top-level concepts that need to be mapped (e.g. `connection`, `stream`, `window`, `priority`) then that's great, and not matching these terms would be confusing. However, terms that just appear (and especially those used inconsistently) in order to express the behavior of a higher-level concept are not necessarily worthy of appearing in the software interfaces. Specifically, the term `resource allocation` I feel is a poor choice for the software interface for 2 reasons:

1.  The term is vague and it's not immediately clear what resources are being allocated. The spec is also purposefully vague about what that term means. The name we choose should indicate that we're specifically setting bytes aside for a stream.
2. The term `allocation` is a name that would lead someone who has not read the RFC to immediately think of `malloc`.  While `assignment` may have its own alternative connotations, they are not as strong to a software engineer as `allocation` would be.

With that said, I'm certainly open to other suggestions :).